### PR TITLE
[WIP][hue] Bypass the light poll during a certain time after a light command

### LINKED ISSUE #7039: [hue] Unexpected Behaviour in Basic UI with Hue Items and auto-update
## Expected Behavior
If i am going to press a switch or a dimmer in the basic ui, the state should change to my decision directly. Also the switch should represent changes of connected devices in the basic ui.

## Current Behavior
The User changes the state of an item in the basic ui. The change is been undo by the openhab server and after the physical item change and the changed state is reported to openhab back the change is then shown in the basic ui. If i am going to describe it in steps, this is how it happens:

0. Base state of item: OFF
1. User wants to turn ON the item
2. User clicks on switch, which switches the item ON
2a. (openhab sends the change to the physical thing)
2b. (openhab detects the current state of the physical item - which is still OFF)
2c. (openhab switches the item back OFF)
3. The physical things switches ON
4. The change is represent in the basic ui and the item is switched ON.

Here is also a shot clip which shows this issue:
https://www.youtube.com/watch?v=r377Pu2vIYw

This also happes for a lot of other sitemap items connected to my hue bridge.

Please keep in mind that the steps 2a through 2c are just educated guesses based on my experience.

## Possible Solution
Don't know

## Steps to Reproduce (for Bugs)
In my opinion this should happen to every hue binding user. If i am wrong, i am will gladly provide my config which is a copy of the examples from the hue binding.

## Context
This is also an annoyance and i am not able to explain this issues to my family.

## Your Environment
I am using the latest openhab docker image from the official openhab repository on docker hub (https://hub.docker.com/r/openhab/openhab/).
By the time of writing this should be version 2.5.1. I am using a philips hue bridge (FW: 1937045000) with standard philips hue e27 bulbs with latest firmware (FW: 1.46.13_r26312) and a hue dimmer switch with current firmware (FW: 6.1.1.28573). The openhab docker image runs on unraid 6.8.2 however i thing this shouldn't matter. The latest chrome version for windows and android. This also happens in the openhab android app so i don't think this is related to the gui.

I also created a topic in the openhab forum about this:
https://community.openhab.org/t/unexpected-sitemap-behaviour-switch-resets-itself-and-gets-updated-a-few-seconds-later-hue-binding-video-included/83478/2

Since october i did a complete fresh install of 2.4 on docker and a complete fresh install of 2.5 in docker. This issue was first seen on a openhab 2.4 installation inside an ubuntu vm on esxi more than 12 months ago.

If it turns out, that this is related to a bug and not a configuration issue, than i will gladly add a bounty to this issue.

### PR DESCRIPTION
**Pull request by lolodomo**
Fix #7039

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by lolodomo** [comment]
Here is a jar for testing:
[org.openhab.binding.hue-2.5.5-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4544430/org.openhab.binding.hue-2.5.5-SNAPSHOT.zip)

**Comment by DerOetzi** [comment]
As I wrote in issue, I would prefer to block only the polling for the light which received the command, not whole polling. For users with lots of lights connected or those with high polling intervals this could lead to very high polling intervals for all lights

**Comment by lolodomo** [comment]
This is doable but more difficult.
Can you at least first test that it fixes the issue ?
If you confirm that the problem is really solved, I will implement the "by light" solution.

**Comment by DerOetzi** [comment]
> 
> 
> This is doable but more difficult.
> Can you at least first test that it fixes the issue ?
> If you confirm that the problem is really solved, I will implement the "by light" solution.

Seems to work for my dimmable lights. Can't reproduce jumping faders.

**Comment by lolodomo** [comment]
If we do at a light level, maybe we should consider a fixed delay + the light fade time, like for example 250 ms + the fade time ?

**Comment by lolodomo** [comment]
Here is a new version which is doing the bypass during 1 second at the light level.
[org.openhab.binding.hue-2.5.5-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4545137/org.openhab.binding.hue-2.5.5-SNAPSHOT.zip)

**Comment by DerOetzi** [comment]
> 
> 
> Here is a new version which is doing the bypass during 1 second at the light level.
> [org.openhab.binding.hue-2.5.5-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4545137/org.openhab.binding.hue-2.5.5-SNAPSHOT.zip)

It seems to work but I'm not happy with the solution to hold the times on bridge level. In my opinion this logic belongs to single lights and therefore should be part of HueLightHandler class. Maybe set timestamp in handleCommand method and check in  onLightStateChanged(@Nullable HueBridge bridge, FullLight fullLight)

**Comment by lolodomo** [comment]
> It seems to work but I'm not happy with the solution to hold the times on bridge level. In my opinion this logic belongs to single lights and therefore should be part of HueLightHandler class. Maybe set timestamp in handleCommand method and check in onLightStateChanged(@nullable HueBridge bridge, FullLight fullLight)

Not possible because the polling job is run by the bridge handler and this job is keeping a cache of the last light states. This job must be able to know if a light state has to be updated in order to updatre or not the value in the cache.

**Comment by DerOetzi** [comment]
> 
> 
> > It seems to work but I'm not happy with the solution to hold the times on bridge level. In my opinion this logic belongs to single lights and therefore should be part of HueLightHandler class. Maybe set timestamp in handleCommand method and check in onLightStateChanged(@nullable HueBridge bridge, FullLight fullLight)
> 
> Not possible because the polling job is run by the bridge handler and this job is keeping a cache of the last light states. This job must be able to know if a light state has to be updated in order to updatre or not the value in the cache.

Ok I see! What is the sense the bridge and not the light or sensor thing is caching states for everyone

**Comment by lolodomo** [comment]
> What is the sense the bridge and not the light or sensor thing is caching states for everyone

The polling job allows a refresh of each created thing + the discovery inbox.
The state cannot be stored in a thing handler of a light because a thing does exist only if the user decides to create one. The light state is also required in the bridge handler to notify the discovery service.
By the way, getting a data from a thing handler in the bridge handler is not easy, you have to go through all the things attached to the bridge thing.

**Comment by DerOetzi** [comment]
> 
> 
> > What is the sense the bridge and not the light or sensor thing is caching states for everyone
> 
> The polling job allows a refresh of each created thing + the discovery inbox.
> The state cannot be stored in a thing handler of a light because a thing and its thing handler does exist only if the user decides to create one. The light state is also required in the bridge handler to notify the discovery service.
> By the way, getting a data from a thing handler in the bridge handler is not easy, you have to go through all the things attached to the bridge thing.

Okay thanks for explanation. One further question on this. What I understand is that bridge thing needs to hold the lastResponse from the hue bridge for discovery and new things. My question do existing things ever request this cached data from bridge later on? Because only existing things can receive commands and therefore have to reject new data. In this case we could move the logic as I suggested to the HueLightHandler, but if existing things although request this cached data from bridge beside getting notified by bridge your solution is the only one.

**Comment by DerOetzi** [comment]
Ok I found the answer to my last question myself in the code. I would suggest a rewrite on this. I can do until this evening and suggest another PR for this, if you are okay with this @lolodomo ?

**Comment by DerOetzi** [comment]
I have made an own suggestion, with refactoring of HueLightHandler, would you have a look @lolodomo:

https://github.com/DerOetzi/openhab-addons/commit/b8174cac6f4dfd1720ceb0b8670f33d4f71a4787

**Comment by J-N-K** [comment]
LGTM, @DerOetzi. But I would like to hear @lolodomo, since he is far more into the code than me.

**Comment by lolodomo** [comment]
@DerOetzi / I see a major problem with your solution. The bridge handler (in the polling job) is calling onLightStateChanged on the light handler only when it detects a change. Imagine you have a poll with a new state for the light, onLightStateChanged is called on the light handler and this light state is igored because arriving too late. At the next poll, as the light state is unchanged, onLightStateChanged on the light handler is not called. As the result, the light state will not be updated.

That being said, I like the other changes in your PR like the suppression of the getLight method.

**Comment by lolodomo** [comment]
@DerOetzi : your solution would work in case the poll job would notify a state changed at every poll. Then the comparison between the old state state and new state should be done in the thing handler. Similar thing should be done by the discovery service. The name "onLightStateChanged " should then be renamed into something like "onNewLightState"

**Comment by lolodomo** [comment]
@DerOetzi : I realized that it would a problem for groups as we have to compute the group state from light states, without using the light thing handlers.

**Comment by DerOetzi** [comment]
> 
> 
> @DerOetzi : your solution would work in case the poll job would notify a state changed at every poll. Then the comparison between the old state state and new state should be done in the thing handler. Similar comparison should be done by the discovery service. The name "onLightStateChanged " should then be renamed into something like "onNewLightState"

@lolodomo thanks for your feedback I can give it a try this afternoon, when I finished my work to do such refactoring. 

Have to  think about a solution for the groups problem, as well.

**Comment by DerOetzi** [comment]
> 
> 
> @DerOetzi : I realized that it would a problem for groups as we have to compute the group state from light states, without using the light thing handlers.

Why is it neccassary to compute without the thing handler? When we register LightStatusHandler we can store them to a Map together with there lightId as key instead of a List like it is at the moment. Then it should be very easy to ask the bridge give me the LightHandler by Id. This would although make it obsolate to iterate and notify each LightHandler, when bridge wants to update a specific light. This would be a little performance optimization as well.

**Comment by lolodomo** [comment]
Because you have no thing hander if you have no thing attached to a light.
And in case there is a light command, it must be known not only by the light handler by the group handler. 
More and more I thing your solution add complexity for no real reaso.

**Comment by lolodomo** [comment]
I just pushed a new version that takes into account the fade time in the light command.
The bypass duration is now 250 ms + the optional fade time.
[org.openhab.binding.hue-2.5.5-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4564011/org.openhab.binding.hue-2.5.5-SNAPSHOT.zip)

**Comment by lolodomo** [comment]
I am not yet satisfied with this version due to the possible enqueueing of several async commands.
The bypass frame should rather be set when the command returns.

**Comment by lolodomo** [comment]
I just pushed a new version that takes into account the fade time in the light command + the possible euqueuing of the commands.
The bypass duration is first set 2 seconds just before sending the light command and then set to 250 ms + the optional transition time when the command returned.

[org.openhab.binding.hue-2.5.5-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4564177/org.openhab.binding.hue-2.5.5-SNAPSHOT.zip)

**Comment by lolodomo** [comment]
A version for testing:
[org.openhab.binding.hue-2.5.5-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4564228/org.openhab.binding.hue-2.5.5-SNAPSHOT.zip)

**Comment by DerOetzi** [comment]
> Because you have no thing hander if you have no thing attached to a light.
> And in case there is a light command, it must be known not only by the light handler by the group handler.
> More and more I thing your solution add complexity for no real reaso.

I'm working on it and think the BridgeHandler is getting less complex now, having same functionality. But I need another day to finish the refactoring.

Here my work in progress:

https://github.com/DerOetzi/openhab-addons/commit/cd4e7e9eccde529c2261aaa5525e73628b6763f9

**Comment by DerOetzi** [comment]
I noticed another thing while doing the refactoring. I think we can start polling jobs right in initialize of bridge and stop it on dispose, because the discovery service always register as listener to both lists and therefore there will be always at least one listener and polling will never stop.

**Comment by lolodomo** [comment]
> I noticed another thing while doing the refactoring. I think we can start polling jobs right in initialize of bridge and stop it on dispose, because the discovery service always register as listener to both lists and therefore there will be always at least one listener and polling will never stop.

Yes, due to the discovery service, the 2 polling jobs will be started. So you are right, it would be simpler to start them immediately.

**Comment by DerOetzi** [comment]
I have started a new PR #7518 with my suggestion for moving the state handling from bridge to thing handler where possible . And refactored the bridge for only notifying the single thing handler and discovery service where needed. Still work in process, but please give me some feedback on the PR what you think. In my opinion this reduces complexity from bridge handler and moves the scope of decision responsibility to the handlers it belongs to.

**Comment by lolodomo** [comment]
I rebased my PR and included the case of the groups.
By thge way, I discovered that the command to change the state of a group should be run asynchronously (and enqueued) like for the lights. I will create a PR for that fix.

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueBridgeHandler.java [hunk 1]
```diff
@@ -265,16 +272,39 @@ public void handleCommand(ChannelUID channelUID, Command command) {
         // not needed
     }
 
+    private void enablePollBypaasBeforeCmd(String lightId) {
+        endBypassFrames.put(lightId, System.currentTimeMillis() + BYPASS_MIN_DURATION_BEFORE_CMD);
+    }
+
+    private void enablePollBypaasAfterCmd(String lightId, @Nullable Long fadeTime) {
```
[the remark this answers is not in the document]
Comment (lolodomo) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueBridgeHandler.java]: Fixed

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueBridgeHandler.java [hunk 2]
```diff
@@ -265,16 +272,39 @@ public void handleCommand(ChannelUID channelUID, Command command) {
         // not needed
     }
 
+    private void enablePollBypaasBeforeCmd(String lightId) {
```
[the remark this answers is not in the document]
Comment (lolodomo) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueBridgeHandler.java]: Fixed

**Comment by lolodomo** [comment]
> Refactor the notification of listeners. Only notify the listeners which are affected not all listeners like it is at the moment. And therefore although separate discovery service from light and sensor listener to make clear, which events are really supported by those things.

To be honest, I am not sure it is worth to do this change.

**Comment by lolodomo** [comment]
> Can you please explain a little bit more to me why bypassing is needed before and after command, in your PR?

@DerOetzi : I am answering your question in the other PR.
The bypass frame set just before the command is just to be sure to have a certain bypass frame, long enough even if several commands are enqueued. I set it to 1.5 seconds, I hope that it is enough.
As soon as the command is run, I update the bypass frame to now only consider the transtition time.

**Comment by lolodomo** [comment]
This PR is covered by PR #7518

**Pull request closed without merging**